src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -2977,37 +2979,27 @@ bool G1PrintRegionLivenessInfoClosure::do_heap_region(HeapRegion* r) {
   _total_remset_bytes    += remset_bytes;
   _total_strong_code_roots_bytes += strong_code_roots_bytes;
 
-  // Print a line for this particular region.
   if(gc_eff < 0) {
-    log_trace(gc, liveness)(G1PPRL_LINE_PREFIX
-                          G1PPRL_TYPE_FORMAT
-                          G1PPRL_ADDR_BASE_FORMAT
-                          G1PPRL_BYTE_FORMAT
-                          G1PPRL_BYTE_FORMAT
-                          G1PPRL_BYTE_FORMAT
-                          G1PPRL_DOUBLE_H_FORMAT
-                          G1PPRL_BYTE_FORMAT
-                          G1PPRL_STATE_FORMAT
-                          G1PPRL_BYTE_FORMAT,
-                          type, p2i(bottom), p2i(end),
-                          used_bytes, prev_live_bytes, next_live_bytes, "-",
-                          remset_bytes, remset_type, strong_code_roots_bytes);
+    snprintf(gc_efficiency, G1PPRL_DOUBLE_FORMAT_LEN+1, G1PPRL_DOUBLE_H_FORMAT, "-");
```

Comment:
snprintf is fine with me too, but I had imagined something like this:

```
FormatBuffer<> efficiency(""); //  maybe better name this
if (gc_eff < 0.0) {
  efficiency.append("-");
} else {
  efficiency.append(G1PPRL_DOUBLE_H_FORMAT, gc_eff);
}
```

and in the `log_trace` use `%s` and `efficiency.buffer()`. That seems a lot easier to understand (for me) than wrangling with `snprintf`.

Maybe there is a reason to not use this?

Also, I am not sure that using `G1PPRL_DOUBLE_H_FORMAT` for this `snprintf` here to print only `-` is really correct. I would have naively have expected to require `%s`.

Reply:
Again, you're absolutely right. 
I looked for a FormatBuffer class in the codebase when you mentioned it (kind of what I wanted from the beginning). For whatever reason I couldn't find one (typo, temporary blindness, ignorance??). So I went with snprintf (even though I don't particular like it). 
I'll fix this. Sorry for all the bother.